File: java-sdk/src/main/java/org/arguslog/sdk/Scrubber.java
package org.arguslog.sdk;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public final class Scrubber {

  private static final String REDACTED = "[Filtered]";

  private static final List<Pattern> DEFAULT =
      List.of(
          Pattern.compile("[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}"),
          Pattern.compile("\\b(?:\\d[ -]*?){13,19}\\b"),
          Pattern.compile("\\beyJ[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+\\.[A-Za-z0-9_./+=-]*\\b"),
          Pattern.compile("\\b(?:\\d{1,3}\\.){3}\\d{1,3}\\b"),
          Pattern.compile("(?i)\\bBearer\\s+[A-Za-z0-9._~+/=-]+"));

  private final List<Pattern> patterns;
  private final boolean enabled;

  public Scrubber(boolean enabled, List<Pattern> extraPatterns) {
    this.enabled = enabled;
    List<Pattern> all = new ArrayList<>(DEFAULT);
    all.addAll(extraPatterns);
    this.patterns = List.copyOf(all);
  }

  public String scrub(String input) {
    if (!enabled || input == null) return input;
    String out = input;
    for (Pattern p : patterns) {
      out = p.matcher(out).replaceAll(REDACTED);
    }
    return out;
  }
}

File: python-sdk/src/arguslog/_scrubber.py
"""PII scrubber. Mirrors the patterns shipped in sdk-core/src/scrubber.ts and
java-sdk/.../Scrubber.java so an event scrubbed in any language looks the same on the wire."""

from __future__ import annotations

import re
from collections.abc import Iterable
from re import Pattern

# Default patterns: email, JWT-shaped tokens, credit-card-ish digit runs.
_DEFAULT_PATTERNS: list[Pattern[str]] = [
    re.compile(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}"),
    re.compile(r"\beyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\b"),
    re.compile(r"\b(?:\d[ -]?){13,19}\b"),
]

REDACTED = "[REDACTED]"


class Scrubber:
    def __init__(self, enabled: bool = True, extra_patterns: Iterable[str] = ()) -> None:
        self._enabled = enabled
        compiled = list(_DEFAULT_PATTERNS)
        for raw in extra_patterns:
            compiled.append(re.compile(raw))
        self._patterns = compiled

    def scrub(self, value: str) -> str:
        if not self._enabled or not value:
            return value
        out = value
        for pattern in self._patterns:
            out = pattern.sub(REDACTED, out)
        return out
